processors/zipkin/src/main/java/org/hawkular/apm/processor/zipkin/CompletionTimeDeriver.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hawkular.apm.processor.zipkin;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.hawkular.apm.api.model.Property;
+import org.hawkular.apm.api.model.events.CompletionTime;
+import org.hawkular.apm.server.api.model.zipkin.Annotation;
+import org.hawkular.apm.server.api.model.zipkin.Span;
+import org.hawkular.apm.server.api.services.SpanCache;
+import org.hawkular.apm.server.api.task.AbstractProcessor;
+import org.hawkular.apm.server.api.task.RetryAttemptException;
+import org.jboss.logging.Logger;
+
+/**
+ * Deriver for completion time of trace reported by zipkin instrumentation.
+ *
+ * @author Pavol Loffay
+ */
+@Singleton
+public class CompletionTimeDeriver extends AbstractProcessor<CompletionTimeProcessing, CompletionTime>{
+
+    private static final Logger log = Logger.getLogger(CompletionTimeDeriver.class.getName());
+
+    private SpanCache spanCache;
+
+
+    public CompletionTimeDeriver() {
+        super(ProcessorType.OneToOne);
+    }
+
+    @Inject
+    public CompletionTimeDeriver(SpanCache spanCache) {
+        this();
+        this.spanCache = spanCache;
+    }
+
+
+    @Override
+    public CompletionTime processOneToOne(String tenantId, CompletionTimeProcessing completionTimeProcessing)
+            throws RetryAttemptException{
+
+        Span rootSpan = completionTimeProcessing.getRootSpan();
+
+        List<Span> trace = spanCache.getTrace(tenantId, rootSpan.getTraceId());
+        Annotation lastAnnotation = extractLastAnnotation(trace);
+
+        /**
+         * TODO HWKAPM-348
+         *
+         * When there is new information recorded in trace (annotation with bigger timestamp)
+         * then wait for a fixed amount of time {@link #getRetryDelay(List, int)} to calculate
+         * trace completion time for all reported spans (async span can be recorded aby time after root
+         * span has been recorded). This approach does not deal with long run async spans,
+         * because it waits only for a fixed amount of time.
+         */
+        if (completionTimeProcessing.getLastTimestamp() == null ||
+                completionTimeProcessing.getLastTimestamp() < lastAnnotation.getTimestamp()) {
+            completionTimeProcessing.setLastTimestamp(lastAnnotation.getTimestamp());
+            return null;
```

Comment:
So just to cleaify - the way this is done when processing the APM trace information:

The processor to initiate the completion information processing: https://github.com/hawkular/hawkular-apm/blob/master/processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionInformationInitiator.java#L86

Then there is the iterative processing, that has TraceCompletionInformation as input and output: https://github.com/hawkular/hawkular-apm/blob/master/processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionInformationProcessor.java#L38 - so this will only do work if there is something to be done.

And then finally the https://github.com/hawkular/hawkular-apm/blob/master/processors/tracecompletiontime-deriver/src/main/java/org/hawkular/apm/processor/tracecompletiontime/TraceCompletionTimeDeriver.java#L31 processor that taks the TraceCompletionInformation and converts it to CompletionTime if there is no further work to do.

So there are two processors subscribing to TraceCompletionInformation, and each event will only essentially be processed by one of the processors, as they are both looking for different conditions (i.e. when to continue processing, or complete and return the CompletionTime).